### PR TITLE
retry automatic announcements of the schema changes on concurrent operation

### LINKED ISSUE #15435: unhandled group0_concurrent_modification exception during concurrent bootstrap
I've been testing concurrent bootstraps that should work in the Raft-based topology (#15423). Sometimes, one of the 3 bootstrapping nodes crashes with
```
ERROR 2023-09-15 16:36:48,131 [shard 0:main] init - Startup failed: service::group0_concurrent_modification (Failed to apply group 0 change due to concurrent modification)
```
The reason is the unhandled `group0_concurrent_modification` exception thrown in `raft_group0_client::add_entry` that is called in `migration_manager::announce_with_raft`.

```
INFO  2023-09-15 16:42:44,674 [shard 0:stre] migration_manager - Create new Keyspace: KSMetaData{name=system_distributed, strategyClass=org.apache.cassandra.locator.SimpleStrategy, strategyOptions={replication_factor=3}, cfMetaData={}, durable_writes=1, userTypes=org.apache.cassandra.config.UTMetaData@0x600004141420}
INFO  2023-09-15 16:42:44,678 [shard 0:comp] compaction - [Compact system.scylla_local 21973b10-53d6-11ee-9b67-27244cebb252] Compacted 2 sstables to [/home/patrykjedrzejczak/forks/scylladb/testlog/dev/scylla-3/data/system/scylla_local-2972ec7ffb2038ddaac1d876f2e3fcbd/mc-14-big-Data.db:level=0]. 63kB to 58kB (~91% of original) in 4ms = 15MB/s. ~256 total partitions merged to 2.
INFO  2023-09-15 16:42:44,755 [shard 0:stre] schema_tables - Creating keyspace system_distributed
INFO  2023-09-15 16:42:44,759 [shard 0:stre] schema_tables - Schema version changed to 8ed4a870-ad4f-3756-b91d-0dab6de3b943
INFO  2023-09-15 16:42:44,760 [shard 0:stre] migration_manager - Create new Keyspace: KSMetaData{name=system_distributed_everywhere, strategyClass=org.apache.cassandra.locator.EverywhereStrategy, strategyOptions={}, cfMetaData={}, durable_writes=1, userTypes=org.apache.cassandra.config.UTMetaData@0x600004141420}
INFO  2023-09-15 16:42:44,856 [shard 0:stre] schema_tables - Creating keyspace system_distributed_everywhere
INFO  2023-09-15 16:42:44,859 [shard 0:stre] schema_tables - Schema version changed to ad8e209a-d451-39fc-9409-5d8975ed86a8
INFO  2023-09-15 16:42:44,860 [shard 0:main] init - Shutting down group 0 service
...
ERROR 2023-09-15 16:36:48,131 [shard 0:main] init - Startup failed: service::group0_concurrent_modification (Failed to apply group 0 change due to concurrent modification)
```

I think the fix is to just catch this exception and retry (like almost everywhere) but I'm not sure. Maybe there is a reason why it's not handled there.

cc @kbr-scylla @gleb-cloudius

### PR DESCRIPTION
**Pull request by patjed41**
The follow-up to #15594.

We retry every automatic `migration_manager::announce` if `group0_concurrent_modification` occurs. Concurrent operations can happen during concurrent bootstrap in Raft-based topology, so we need this change to enable support for concurrent bootstrap.

This PR adds retry loops in 4 places:
- `service::create_keyspace_if_missing`,
- `system_distributed_keyspace::start`,
- `redis::create_keyspace_if_not_exists_impl`,
- `table_helper::setup_keyspace` (used for creating the `system_traces` keyspace).

Fixes #15435

### CODE REVIEW

**Review by kbr-scylla — APPROVED** [review]
Last 7 commits LGTM.

File: table_helper.cc
Comment (piodul) [inline table_helper.cc:186]: Ditto

**Review by piodul — CHANGES_REQUESTED** [review]

**Comment by patjed41** [comment]
v2:
- flipped the condition at the end of `create_keyspace_if_not_exists_impl` from `!mutations.empty()` to `mutations.empty()`
- rebased after v3 of #15594

File: redis/keyspace_utils.cc
Comment (patjed41) [reply redis/keyspace_utils.cc]: It should, I've fixed it.

File: table_helper.cc
Comment (patjed41) [reply table_helper.cc:157]: I was following the logging practices in those places. `table_helper` doesn't log anything even in exceptional situations. It doesn't even have a logger. On the other hand, other places log information in similar situations.

**Comment by patjed41** [comment]
> CI state FAILURE - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4048/

Timeouts in gating on ALTER and TRUNCATE that has been appearing recently in many tests.

**Comment by patjed41** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4112/

known flaky test -- #15452

**Comment by patjed41** [comment]
rebased after v4 of #15594

**Comment by patjed41** [comment]
rebased on the new #15594

**Comment by kbr-scylla** [comment]
> [2023-10-16T14:11:15.376Z] clang-16: error: unable to execute command: Bus error

It could be some problem with the machine.
In these cases I simply restart CI.

**Comment by kbr-scylla** [comment]
Please rebase

**Comment by patjed41** [comment]
I rebased and removed the follow-up info from the description.

File: table_helper.cc
Comment (piodul) [reply table_helper.cc:157]: I see. I think we should go ahead and add a logger, then log a message about retry here like in other places. While in general we should eventually stop getting the `group0_concurrent_modification` exception, I think we might theoretically get stuck here due to a livelock because another node keeps doing group0 operations faster than us - having this message would let us know where we got stuck.

**Review by gleb-cloudius — APPROVED** [review]

**Comment by patjed41** [comment]
v3:
- added a logger to `table_helper.cc`
- added logging info about concurrent operations in `table_helper::setup_keyspace`

**Review by piodul — APPROVED** [review]

**Review by kbr-scylla — APPROVED** [review]

**Comment by kbr-scylla** [comment]
Dequeued the PR
According to @michoecho the `table_helper: retry in setup_keyspace on concurrent operation` commit uncovers a bug in `coroutine::parallel_for_each`.

@patjed41 to unblock the PR we can switch to regular `parallel_for_each`, we're not on a hot path anyway.

**Comment by gleb-cloudius** [comment]
> Dequeued the PR According to @michoecho the `table_helper: retry in setup_keyspace on concurrent operation` commit uncovers a bug in `coroutine::parallel_for_each`.
> 

What bug?

**Comment by michoecho** [comment]
> > Dequeued the PR According to @michoecho the `table_helper: retry in setup_keyspace on concurrent operation` commit uncovers a bug in `coroutine::parallel_for_each`.
> 
> What bug?

https://github.com/scylladb/seastar/issues/1902

**Comment by gleb-cloudius** [comment]
> > > Dequeued the PR According to @michoecho the `table_helper: retry in setup_keyspace on concurrent operation` commit uncovers a bug in `coroutine::parallel_for_each`.
> > 
> > 
> > What bug?
> 
> [scylladb/seastar#1902](https://github.com/scylladb/seastar/issues/1902)

So why this patch suddenly uncovers the bug given that the `coroutine::parallel_for_each` use was there before the patch as well?

**Comment by michoecho** [comment]
> > > > Dequeued the PR According to @michoecho the `table_helper: retry in setup_keyspace on concurrent operation` commit uncovers a bug in `coroutine::parallel_for_each`.
> > > 
> > > 
> > > What bug?
> > 
> > 
> > [scylladb/seastar#1902](https://github.com/scylladb/seastar/issues/1902)
> 
> So why this patch suddenly uncovers the bug given that the `coroutine::parallel_for_each` use was there before the patch as well?

I don't know and I don't think it's worth understanding in-depth.

**Comment by gleb-cloudius** [comment]
> > > > > Dequeued the PR According to @michoecho the `table_helper: retry in setup_keyspace on concurrent operation` commit uncovers a bug in `coroutine::parallel_for_each`.
> > > > 
> > > > 
> > > > What bug?
> > > 
> > > 
> > > [scylladb/seastar#1902](https://github.com/scylladb/seastar/issues/1902)
> > 
> > 
> > So why this patch suddenly uncovers the bug given that the `coroutine::parallel_for_each` use was there before the patch as well?
> 
> I don't know and I don't think it's worth understanding in-depth.

We have a lot of code in production that uses `coroutine::parallel_for_each`. It will be interesting to know how vulnerable is it.

**Comment by michoecho** [comment]
> We have a lot of code in production that uses coroutine::parallel_for_each. It will be interesting to know how vulnerable is it.

The conditions for the bug to manifest are:
1. The `coroutine::parallel_for_each` was suspended (i.e. didn't finish immediately).
2. After the `coroutine::parallel_for_each` finishes, but before control returns to the reactor, `current_tasktrace()` is called.
3. The used-after-free memory has been overwritten at the time of the `current_tasktrace()` call.

Apparently before the PR some of these conditions didn't occur. But figuring out which exactly it was would require a chunk of work, and it wouldn't tell you anything about other uses of `coroutine::parallel_for_each`.

Also, this is the kind of undefined behaviour which pretty much can only result in a segfault, and not something more insidious. So we would have noticed if it was happening in production.

Calls to `current_tasktrace()` are very rare. In this particular dtest, they are happening only because TRACE logs are enabled.

**Comment by patjed41** [comment]
I'll rebase after merging #15897.

**Comment by patjed41** [comment]
rebased

**Pull request merged**
